contracts/utils/convex/ConvexStakingWrapperYield.sol

```diff
@@ -0,0 +1,112 @@
+import '@yield-protocol/vault-interfaces/ICauldron.sol';
+import '@yield-protocol/vault-interfaces/DataTypes.sol';
+import './ConvexStakingWrapper.sol';
+
+/// @title Convex staking wrapper for Yield platform
+/// @notice Enables use of convex LP positions as collateral while still receiving rewards
+contract ConvexStakingWrapperYield is ConvexStakingWrapper {
+    using TransferHelper for IERC20;
+
+    /// @notice Mapping to keep track of the user & their vaults
+    mapping(address => bytes12[]) public vaults;
+
+    ICauldron public cauldron;
+
+    /// @notice Event called when a vault is set for a user
+    /// @param account The account for which vault is set
+    /// @param vault The vaultId
+    event VaultSet(address indexed account, bytes12 indexed vault);
+
+    constructor(
+        address curveToken_,
+        address convexToken_,
+        address convexPool_,
+        uint256 poolId_,
+        address join_,
+        ICauldron cauldron_
+    ) {
+        name = string(abi.encodePacked('Wrapped ', ERC20(convexToken_).name()));
+        symbol = string(abi.encodePacked('wrp', ERC20(convexToken_).symbol()));
+        isShutdown = false;
+        isInit = true;
+        curveToken = curveToken_;
+        convexToken = convexToken_;
+        convexPool = convexPool_;
+        convexPoolId = poolId_;
+        collateralVault = join_; //TODO: Add the join address
+        cauldron = cauldron_;
+    }
+
+    /// @notice Points the collateral vault to the join storing the wrappedCvx3Crv
+    /// @param join_ Join which will store the wrappedCvx3Crv of the user
+    function point(address join_) external auth {
+        collateralVault = join_;
+    }
+
+    /// @notice Adds a vault to the user's vault list
+    /// @param vaultId The vaulId being added
+    function addVault(bytes12 vaultId) external {
+        address account = cauldron.vaults(vaultId).owner;
+        require(account != address(0), 'No owner for the vault');
+        bytes12[] storage vaults_ = vaults[account];
+        for (uint256 i = 0; i < vaults_.length; i++) {
+            require(vaults_[i] != vaultId, 'already added');
+        }
+        vaults_.push(vaultId);
+        vaults[account] = vaults_;
+        emit VaultSet(account, vaultId);
+    }
+
+    /// @notice Get user's balance of collateral deposited in various vaults
+    /// @param account_ User's address for which balance is requested
+    /// @return User's balance of collateral
+    function _getDepositedBalance(address account_) internal view override returns (uint256) {
+        if (account_ == address(0) || account_ == collateralVault) {
+            return 0;
+        }
+
+        if (vaults[account_].length == 0) {
+            return _balanceOf[account_];
+        }
+        bytes12[] memory userVault = vaults[account_];
+
+        //add up all balances of all vaults
+        uint256 collateral;
+        DataTypes.Balances memory balance;
+        for (uint256 i = 0; i < userVault.length; i++) {
```

Comment:
Here we need to check that `account_` is still the owner of `userVault[i]`, and call `removeVault` if it is not. We then also need to ignore vaultIds that are zero.

Reply:
Do we want to do this as it will lead to the getDepositedBalance no longer being a view function which would add some more gas costs everytime user calls the getReward function. Also here are more functions which could no longer be view functions:

1. earned
2. totalBalanceOf

Reply:
`_getDepositedBalance` is an internal function that is only called by mutable (non-view) functions in ConvexYieldWrapper.sol. It makes no difference for it to be `view` or not (`view` functions cost gas when called from mutable functions).

Reply:
The alternative is to leave it as a `view` function, but check the `owner` of each vault before adding its collateral balance. That's an extra `CALL` per vault. It's probably better than making `earned` and `totalBalanceOf` mutable.